### ISSUE TITLE
Reporting of PRETRANSFER_TIME and STARTTRANSFER_TIME when using curl_multi_perform()

### ISSUE THREAD

**Issue by lonezor** [curl/curl#4776]
### I did this
I'm seeing this in libcurl-7.67.0 using HTTP/2 backend when sending POST in
custom build environment, but it can also be seen with HTTP 1.1 GET so
I'm reproducing it with Ubuntu package libcurl3-gnutls:amd64.

The example multi-single.c has been modified to run a larger number
of transfers to localhost's default Ubuntu page (12 kB HTML)

To make it more readable I'm exaggerating the test by running 3000
transfers over two re-used connections:

```
0001 code 200 name_lookup 0.010411 connect 0.010475 app_connect 0.000000 pre_transfer 0.010502 start_transfer 0.011029 total_time 0.011038
0002 code 200 name_lookup 0.010935 connect 0.010957 app_connect 0.000000 pre_transfer 0.010974 start_transfer 0.018878 total_time 0.019320
0003 code 200 name_lookup 0.000012 connect 0.000012 app_connect 0.000012 pre_transfer 0.000043 start_transfer 0.008355 total_time 0.019235
0004 code 200 name_lookup 0.000005 connect 0.000005 app_connect 0.000005 pre_transfer 0.000022 start_transfer 0.008036 total_time 0.027650
0005 code 200 name_lookup 0.000004 connect 0.000004 app_connect 0.000004 pre_transfer 0.000012 start_transfer 0.008417 total_time 0.027657
...
0500 code 200 name_lookup 0.000005 connect 0.000005 app_connect 0.000005 pre_transfer 0.000020 start_transfer 0.006393 total_time 1.919017
0501 code 200 name_lookup 0.000021 connect 0.000021 app_connect 0.000021 pre_transfer 0.000028 start_transfer 0.006649 total_time 1.919024
0502 code 200 name_lookup 0.000004 connect 0.000004 app_connect 0.000004 pre_transfer 0.000025 start_transfer 0.006327 total_time 1.925707
0503 code 200 name_lookup 0.000021 connect 0.000021 app_connect 0.000021 pre_transfer 0.000028 start_transfer 0.006667 total_time 1.925714
0504 code 200 name_lookup 0.000004 connect 0.000005 app_connect 0.000005 pre_transfer 0.000016 start_transfer 0.006313 total_time 1.932292
0505 code 200 name_lookup 0.000020 connect 0.000020 app_connect 0.000020 pre_transfer 0.000027 start_transfer 0.006567 total_time 1.932297
...
2995 code 200 name_lookup 0.000003 connect 0.000003 app_connect 0.000003 pre_transfer 0.000009 start_transfer 0.000195 total_time 6.155638
2996 code 200 name_lookup 0.000003 connect 0.000003 app_connect 0.000003 pre_transfer 0.000010 start_transfer 0.000204 total_time 6.155843
2997 code 200 name_lookup 0.000003 connect 0.000003 app_connect 0.000003 pre_transfer 0.000008 start_transfer 0.000201 total_time 6.155846
2998 code 200 name_lookup 0.000004 connect 0.000004 app_connect 0.000004 pre_transfer 0.000014 start_transfer 0.000195 total_time 6.156043
2999 code 200 name_lookup 0.000003 connect 0.000003 app_connect 0.000003 pre_transfer 0.000009 start_transfer 0.000188 total_time 6.156047
3000 code 200 name_lookup 0.000003 connect 0.000003 app_connect 0.000003 pre_transfer 0.000043 start_transfer 0.000228 total_time 6.156275

real    0m6.220s
user    0m6.145s
sys     0m0.060s
```

As seen above, we can see that server name resolution, connect and app_connect
are done twice and then re-used. This looks fine and just to clarify we are not
using SSL so the value is therefore zero for app_connect. When retrieved from
the cache this is indicated with a few microseconds.

Now, looking at **pre_transfer** and **start_transfer**. Since we are running 3000
transfers I assume that not all of them can be sending bytes on the wire
at the same time and they would all start at different times.

The values listed for all but the first two transfers have a value that in
the timeline is before the original connections started, e.g. pre_transfer
0.000043 and start_transfer 0.008355, when start_transfer at earliest
began 0.011029 for the first connection.

This is what is a bit confusing. Also looking at total_time which covers
the duration of the whole transfer, it currently describes an increasing
time diff between start_transfer and total_time for each new transfer
we add to the test.

Example, the last transfer 6.156275 - 0.000228 = 6.156047 is a very long
time and cannot be correct since the whole program ran for roughly the
same duration with 3000 transfers.

https://curl.haxx.se/libcurl/c/CURLINFO_PRETRANSFER_TIME.html

"Get the time until the file transfer start"

https://curl.haxx.se/libcurl/c/CURLINFO_STARTTRANSFER_TIME.html

"Get the time until the first byte is received"

Question:
Are pre_transfer and start_transfer only applicable for transfers that are
not re-using a connection?
(the values for these two lines above looks ok)

Conclusion:
I think there is a need to have useful values for pre_transfer and
start_transfer time measurements for all types of transfers (supporting all HTTP versions).

### curl/libcurl version
curl 7.58.0 (x86_64-pc-linux-gnu) libcurl/7.58.0 OpenSSL/1.1.1 zlib/1.2.11 libidn2/2.0.4 libpsl/0.19.1 (+libidn2/2.0.4) nghttp2/1.30.0 librtmp/2.3
Release-Date: 2018-01-24
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL

### operating system
Linux johan-LDT-703-1006 4.15.0-72-generic #81-Ubuntu SMP Tue Nov 26 12:20:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

### example program
```
/***************************************************************************
 *                                  _   _ ____  _
 *  Project                     ___| | | |  _ \| |
 *                             / __| | | | |_) | |
 *                            | (__| |_| |  _ <| |___
 *                             \___|\___/|_| \_\_____|
 *
 * Copyright (C) 1998 - 2018, Daniel Stenberg, <daniel@haxx.se>, et al.
 *
 * This software is licensed as described in the file COPYING, which
 * you should have received as part of this distribution. The terms
 * are also available at https://curl.haxx.se/docs/copyright.html.
 *
 * You may opt to use, copy, modify, merge, publish, distribute and/or sell
 * copies of the Software, and permit persons to whom the Software is
 * furnished to do so, under the terms of the COPYING file.
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
 * KIND, either express or implied.
 *
 ***************************************************************************/ 
/* <DESC>
 * Modified to show time measurements
 * </DESC>
 */ 
 
#include <stdio.h>
#include <string.h>
 
/* somewhat unix-specific */ 
#include <sys/time.h>
#include <unistd.h>
 
/* curl stuff */ 
#include <curl/curl.h>
 
#ifdef _WIN32
#define WAITMS(x) Sleep(x)
#else
/* Portable sleep for platforms other than Windows. */ 
#define WAITMS(x)                               \
  struct timeval wait = { 0, (x) * 1000 };      \
  (void)select(0, NULL, NULL, NULL, &wait);
#endif

#define NR_TRANSFERS (3000)
#define CONCURRENCY (2L)

static size_t
write_cb(void* contents, size_t size, size_t nmemb, void* user_data)
{
    return size * nmemb;
}

int main(void)
{
  CURL *http_handle;
  CURLM *multi_handle;

  curl_global_init(CURL_GLOBAL_DEFAULT);

  /* init a multi stack */ 
  multi_handle = curl_multi_init();

  curl_multi_setopt(multi_handle, CURLMOPT_MAX_TOTAL_CONNECTIONS, CONCURRENCY);

  int i;
  for(i=0; i < NR_TRANSFERS; i++) {
    http_handle = curl_easy_init();

    curl_easy_setopt(http_handle, CURLOPT_URL, "http://localhost");
    curl_easy_setopt(http_handle, CURLOPT_WRITEFUNCTION, write_cb);

    /* add the individual transfers */ 
    curl_multi_add_handle(multi_handle, http_handle);
  }

  /* we start some action by calling perform right away */ 
  int still_running = 0; /* keep number of running handles */ 
  int repeats = 0;

  printf("Running curl_multi_perform with %d transfers over %ld connections\n", NR_TRANSFERS, CONCURRENCY);

  curl_multi_perform(multi_handle, &still_running);

  while(still_running) {
    CURLMcode mc; /* curl_multi_wait() return code */ 
    int numfds;
 
    /* wait for activity, timeout or "nothing" */ 
    mc = curl_multi_wait(multi_handle, NULL, 0, 1000, &numfds);
 
    if(mc != CURLM_OK) {
      fprintf(stderr, "curl_multi_wait() failed, code %d.\n", mc);
      break;
    }
 
    /* 'numfds' being zero means either a timeout or no file descriptors to
       wait for. Try timeout on first occurrence, then assume no file
       descriptors and no file descriptors to wait for means wait for 100
       milliseconds. */ 
 
    if(!numfds) {
      repeats++; /* count number of repeated zero numfds */ 
      if(repeats > 1) {
        WAITMS(100); /* sleep 100 milliseconds */ 
      }
    }
    else
      repeats = 0;
 
    curl_multi_perform(multi_handle, &still_running);
  }

  // Print stats when finished
  struct CURLMsg *m = NULL;
  int counter = 1;
  do {
    int msgq = 0;
    m = curl_multi_info_read(multi_handle, &msgq);
    if (m && (m->msg == CURLMSG_DONE)) {
      CURL *e = m->easy_handle;
      long response_code = 0;
      double name_lookup_time = 0;
      double connect_time = 0;
      double app_connect_time = 0;
      double tcp_connect_time = 0;
      double ssl_connect_time = 0;
      double pre_transfer_time = 0;
      double start_transfer_time = 0;
      double total_time = 0;

      curl_easy_getinfo(e, CURLINFO_RESPONSE_CODE, &response_code);
      curl_easy_getinfo(e, CURLINFO_NAMELOOKUP_TIME, &name_lookup_time);
      curl_easy_getinfo(e, CURLINFO_CONNECT_TIME, &connect_time);
      curl_easy_getinfo(e, CURLINFO_APPCONNECT_TIME, &app_connect_time);
      curl_easy_getinfo(e, CURLINFO_PRETRANSFER_TIME, &pre_transfer_time);
      curl_easy_getinfo(e, CURLINFO_STARTTRANSFER_TIME, &start_transfer_time);
      curl_easy_getinfo(e, CURLINFO_TOTAL_TIME, &total_time);

      printf("%04d code %ld name_lookup %f connect %f app_connect %f pre_transfer %f start_transfer %f total_time %f\n", counter++,
             response_code, name_lookup_time, connect_time, app_connect_time, pre_transfer_time, start_transfer_time, total_time);
    }
  } while(m);
 
  curl_multi_remove_handle(multi_handle, http_handle);
 
  curl_easy_cleanup(http_handle);
 
  curl_multi_cleanup(multi_handle);
 
  curl_global_cleanup();
 
  return 0;
}

```

**Comment by bagder** [comment]
> Are pre_transfer and start_transfer only applicable for transfers that are not re-using a connection? (the values for these two lines above looks ok)

No. There can be more and other waiting times than just waiting for the name resolve and TCP connection setup. Like for example limiting the parallelism per host name or total number of connections.

> I think there is a need to have useful values for pre_transfer and start_transfer time measurements for all types of transfers (supporting all HTTP versions).

They should work independently of HTTP versions and connection reuse, yes. If you can reproduce a problem with a current libcurl version I'd be interested in looking into it. 

> Example, the last transfer 6.156275 - 0.000228 = 6.156047 is a very long time and cannot be correct

Isn't that happening because you limit the concurrency so that last transfer had to wait for quite a long time?

**Comment by lonezor** [comment]
> Isn't that happening because you limit the concurrency so that last transfer had to wait for quite a long time?

yes, with limited concurrency and a very large number of transfers we are forcing most of them to wait. This is only done to highlight the issue, we can use less transfers as well but the values would be closer to each other.

But if STARTTRANSFER_TIME indicated when in time the transfer actually started, e.g. 6.100001 instead of  0.000228, then 6.156275 - 6.100001 = 0.056274 and we can understand how much network time is involved.

Having some kind of measurement in Curl in this context of the round-trip-time for e.g.  a request/response between client and server is useful information.

Isn't the example above with a value similar to 6.100001 rather than 0.000228 above more in line with what STARTTRANSFER_TIME should describe? (and similar for PRETRANSFER_TIME)

> If you can reproduce a problem with a current libcurl version I'd be interested in looking into it.

I have reproduced the same thing step-by-step with latest release:
```
lonezor@howaborm:/tmp$ mkdir curl_test

lonezor@howaborm:/tmp$ cd curl_test

lonezor@howaborm:/tmp/curl_test$ curl https://curl.haxx.se/download/curl-7.67.0.tar.gz -o curl-7.67.0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 3975k  100 3975k    0     0  1917k      0  0:00:02  0:00:02 --:--:-- 1917k

lonezor@howaborm:/tmp/curl_test$ mkdir root

lonezor@howaborm:/tmp/curl_test$ tar zxvf curl-7.67.0.tar.gz
curl-7.67.0/
curl-7.67.0/projects/
curl-7.67.0/projects/README
curl-7.67.0/projects/generate.bat
curl-7.67.0/projects/wolfssl_override.props
curl-7.67.0/projects/checksrc.bat
curl-7.67.0/projects/wolfssl_options.h
curl-7.67.0/projects/build-openssl.bat
...
configure: Configured to build curl/libcurl:

  Host setup:       x86_64-pc-linux-gnu
  Install prefix:   /tmp/curl_test/root
  Compiler:         gcc
   CFLAGS:          -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread
   CPPFLAGS:        
   LDFLAGS:         
   LIBS:            -lz

  curl version:     7.67.0
  SSL:              no      (--with-{ssl,gnutls,nss,mbedtls,wolfssl,schannel,secure-transport,mesalink,amissl} )
  SSH:              no      (--with-libssh2)
  zlib:             enabled
  brotli:           no      (--with-brotli)
  GSS-API:          no      (--with-gssapi)
  TLS-SRP:          no      (--enable-tls-srp)
  resolver:         POSIX threaded
  IPv6:             enabled
  Unix sockets:     enabled
  IDN:              no      (--with-{libidn2,winidn})
  Build libcurl:    Shared=yes, Static=yes
  Built-in manual:  enabled
  --libcurl option: enabled (--disable-libcurl-option)
  Verbose errors:   enabled (--disable-verbose)
  Code coverage:    disabled
  SSPI:             no      (--enable-sspi)
  ca cert bundle:   no
  ca cert path:     
  ca fallback:      
  LDAP:             no      (--enable-ldap / --with-ldap-lib / --with-lber-lib)
  LDAPS:            no      (--enable-ldaps)
  RTSP:             enabled
  RTMP:             no      (--with-librtmp)
  Metalink:         no      (--with-libmetalink)
  PSL:              no      (libpsl not found)
  Alt-svc:          no      (--enable-alt-svc)
  HTTP2:            disabled (--with-nghttp2)
  HTTP3:            disabled (--with-ngtcp2, --with-quiche)
  ESNI:             no      (--enable-esni)
  Protocols:        DICT FILE FTP GOPHER HTTP IMAP POP3 RTSP SMTP TELNET TFTP
  Features:         IPv6 UnixSockets libz AsynchDNS

lonezor@howaborm:/tmp/curl_test/curl-7.67.0$ make -j 16 install
Making install in lib
make[1]: Entering directory '/tmp/curl_test/curl-7.67.0/lib'
  CC       libcurl_la-file.lo
  CC       libcurl_la-timeval.lo
  CC       libcurl_la-base64.lo
 ...
 /usr/bin/install -c -m 644 libcurl.m4 '/tmp/curl_test/root/share/aclocal'
 /usr/bin/install -c -m 644 curl_easy_cleanup.3 curl_easy_duphandle.3 curl_easy_escape.3 curl_easy_getinfo.3 curl_easy_init.3 curl_easy_pause.3 curl_easy_perform.3 curl_easy_recv.3 curl_easy_reset.3 curl_easy_send.3 curl_easy_setopt.3 curl_easy_strerror.3 curl_easy_unescape.3 curl_easy_upkeep.3 curl_escape.3 curl_formadd.3 curl_formfree.3 curl_formget.3 curl_free.3 curl_getdate.3 curl_getenv.3 curl_global_cleanup.3 curl_global_init.3 curl_global_init_mem.3 curl_global_sslset.3 curl_mime_addpart.3 curl_mime_data.3 curl_mime_data_cb.3 curl_mime_encoder.3 curl_mime_filedata.3 curl_mime_filename.3 curl_mime_free.3 curl_mime_headers.3 curl_mime_init.3 curl_mime_name.3 curl_mime_subparts.3 curl_mime_type.3 curl_mprintf.3 curl_multi_add_handle.3 curl_multi_assign.3 '/tmp/curl_test/root/share/man/man3'
 /usr/bin/install -c -m 644 curl_multi_cleanup.3 curl_multi_fdset.3 curl_multi_info_read.3 curl_multi_init.3 curl_multi_perform.3 curl_multi_poll.3 curl_multi_remove_handle.3 curl_multi_setopt.3 curl_multi_socket.3 curl_multi_socket_action.3 curl_multi_socket_all.3 curl_multi_strerror.3 curl_multi_timeout.3 curl_multi_wait.3 curl_share_cleanup.3 curl_share_init.3 curl_share_setopt.3 curl_share_strerror.3 curl_slist_append.3 curl_slist_free_all.3 curl_strequal.3 curl_strnequal.3 curl_unescape.3 curl_url.3 curl_url_cleanup.3 curl_url_dup.3 curl_url_get.3 curl_url_set.3 curl_version.3 curl_version_info.3 libcurl-easy.3 libcurl-env.3 libcurl-errors.3 libcurl-multi.3 libcurl-security.3 libcurl-share.3 libcurl-symbols.3 libcurl-thread.3 libcurl-tutorial.3 libcurl-url.3 '/tmp/curl_test/root/share/man/man3'
 /usr/bin/install -c -m 644 libcurl.3 '/tmp/curl_test/root/share/man/man3'
make[6]: Leaving directory '/tmp/curl_test/curl-7.67.0/docs/libcurl'
make[5]: Leaving directory '/tmp/curl_test/curl-7.67.0/docs/libcurl'
make[4]: Leaving directory '/tmp/curl_test/curl-7.67.0/docs/libcurl'
make[3]: Leaving directory '/tmp/curl_test/curl-7.67.0'
make[2]: Leaving directory '/tmp/curl_test/curl-7.67.0'
make[1]: Leaving directory '/tmp/curl_test/curl-7.67.0'

lonezor@howaborm:/tmp/curl_test/curl-7.67.0$ cd ../root/

# example code but with 10k transfers since this machine is faster
lonezor@howaborm:/tmp/curl_test/root$ nano test.c 

lonezor@howaborm:/tmp/curl_test/root$ export LD_LIBRARY_PATH=/tmp/curl_test/root/lib

lonezor@howaborm:/tmp/curl_test/root$ gcc test.c -I /tmp/curl_test/root/include -o test -L /tmp/curl_test/root/lib -lcurl

lonezor@howaborm:/tmp/curl_test/root$ ldd test
	linux-vdso.so.1 (0x00007ffec31af000)
	libcurl.so.4 => /tmp/curl_test/root/lib/libcurl.so.4 (0x00007f39fc03f000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f39fbe29000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f39fbe0d000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f39fbdea000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f39fc0ab000)

lonezor@howaborm:/tmp/curl_test/root$ strings /tmp/curl_test/root/lib/libcurl.so.4 | grep "CLIENT libcurl"
CLIENT libcurl 7.67.0
CLIENT libcurl 7.67.0
CLIENT libcurl 7.67.0

lonezor@howaborm:/tmp/curl_test/root$ time ./test

Running curl_multi_perform with 10000 transfers over 2 connections
0001 code 200 name_lookup 0.042769 connect 0.042837 app_connect 0.000000 pre_transfer 0.042874 start_transfer 0.044311 total_time 0.044323
0002 code 200 name_lookup 0.042777 connect 0.042796 app_connect 0.000000 pre_transfer 0.042810 start_transfer 0.044229 total_time 0.044242
0003 code 200 name_lookup 0.000014 connect 0.000014 app_connect 0.000000 pre_transfer 0.000030 start_transfer 0.001223 total_time 0.045441
0004 code 200 name_lookup 0.000006 connect 0.000006 app_connect 0.000000 pre_transfer 0.000015 start_transfer 0.001214 total_time 0.045450
0005 code 200 name_lookup 0.000007 connect 0.000007 app_connect 0.000000 pre_transfer 0.000024 start_transfer 0.001108 total_time 0.046549
0006 code 200 name_lookup 0.000006 connect 0.000006 app_connect 0.000000 pre_transfer 0.000016 start_transfer 0.001095 total_time 0.046557
0007 code 200 name_lookup 0.000007 connect 0.000007 app_connect 0.000000 pre_transfer 0.000023 start_transfer 0.001074 total_time 0.047622
0008 code 200 name_lookup 0.000007 connect 0.000007 app_connect 0.000000 pre_transfer 0.000021 start_transfer 0.001062 total_time 0.047629
0009 code 200 name_lookup 0.000007 connect 0.000008 app_connect 0.000000 pre_transfer 0.000023 start_transfer 0.001069 total_time 0.048690
0010 code 200 name_lookup 0.000006 connect 0.000006 app_connect 0.000000 pre_transfer 0.000015 start_transfer 0.001057 total_time 0.048695
...
9990 code 200 name_lookup 0.000007 connect 0.000007 app_connect 0.000000 pre_transfer 0.000014 start_transfer 0.000900 total_time 4.942967
9991 code 200 name_lookup 0.000007 connect 0.000007 app_connect 0.000000 pre_transfer 0.000034 start_transfer 0.000951 total_time 4.970789
9992 code 200 name_lookup 0.000007 connect 0.000007 app_connect 0.000000 pre_transfer 0.000015 start_transfer 0.000930 total_time 4.970799
9993 code 200 name_lookup 0.000009 connect 0.000009 app_connect 0.000000 pre_transfer 0.000021 start_transfer 0.000918 total_time 4.957020
9994 code 200 name_lookup 0.000006 connect 0.000006 app_connect 0.000000 pre_transfer 0.000013 start_transfer 0.000908 total_time 4.957027
9995 code 200 name_lookup 0.000009 connect 0.000009 app_connect 0.000000 pre_transfer 0.000021 start_transfer 0.000920 total_time 4.942458
9996 code 200 name_lookup 0.000006 connect 0.000006 app_connect 0.000000 pre_transfer 0.000013 start_transfer 0.000909 total_time 4.942464
9997 code 200 name_lookup 0.000008 connect 0.000008 app_connect 0.000000 pre_transfer 0.000026 start_transfer 0.000936 total_time 4.951804
9998 code 200 name_lookup 0.000007 connect 0.000007 app_connect 0.000000 pre_transfer 0.000014 start_transfer 0.000924 total_time 4.951812
9999 code 200 name_lookup 0.000007 connect 0.000007 app_connect 0.000000 pre_transfer 0.000021 start_transfer 0.000893 total_time 4.968314
10000 code 200 name_lookup 0.000005 connect 0.000005 app_connect 0.000000 pre_transfer 0.000012 start_transfer 0.000882 total_time 4.968321

real	0m5.159s
user	0m4.869s
sys	0m0.217s
```

**Comment by RedDwarf69** [comment]
I would say the problem arises from the fact curl handles a lot of protocols with an uniform interface.

Reading https://curl.haxx.se/libcurl/c/curl_easy_getinfo.html you get a general idea of what NAMELOOKUP, CONNECT, APPCONNECT, PRETRANSFER, STARTTRANSFER and TOTAL report. But when trying to map the descriptions in there to the details of a specific protocol things are not that clear. For example, after using this patch

```
@@ -70,11 +70,6 @@
 
     curl_easy_setopt(http_handle, CURLOPT_URL, "http://localhost");
     curl_easy_setopt(http_handle, CURLOPT_WRITEFUNCTION, write_cb);
-#ifndef PRIOR_KNOWLEDGE
-    curl_easy_setopt(http_handle, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2_0);
-#else
-    curl_easy_setopt(http_handle, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2_PRIOR_KNOWLEDGE);
-#endif
 
     /* add the individual transfers */ 
     curl_multi_add_handle(multi_handle, http_handle);
```

The version with PRIOR_KNOWLEDGE reports
`3000 code 200 name_lookup 0.000000 connect 0.000000 app_connect 0.000000 pre_transfer 0.000073 start_transfer 0.957867 total_time 0.990526`

the version without
`3000 code 0 name_lookup 0.000005 connect 0.000038 app_connect 0.000000 pre_transfer 0.000053 start_transfer 0.004450 total_time 6.585256`

Are those values wrong?

Well, STARTTRANSFER is supposed to report `The time it took from the start until the first byte is received by libcurl`. And if by "first byte" you mean the first byte from the HTTP 1.1 upgrade request (done only once for all further HTTP 2.0 requests) then the difference in the reported start_transfer values makes sense. But:

a) AFAICS the documentation is not 100% clear about whether "first byte" means the reply to the HTTP 1.1 upgrade request or to the HTTP2 request (and depending on how you look at it any would make sense)
b) You may well want to know both. How do you do that? Adding more CURLINFO_*_TIME variables... which while theoretically generic, they may end up being HTTP2 specific forever?


So at the very least it would be nice to have documentation about how the CURLINFO_*_TIME definitions translate to specific protocols when this is not evident.
I don't have a clear opinion about whether the current meaning should be changed (someone may be depending on it), new variables are needed or something else.

**Comment by bagder** [comment]
> Isn't the example above with a value similar to 6.100001 rather than 0.000228 above more in line with what STARTTRANSFER_TIME should describe?

Yes I would agree, but that's not what the code currently does. `TIMER_STARTSINGLE`  is set when it first *considers* doing the connect and will be stored even if it at that point finds out it needs to queue it and perform it later:

https://github.com/curl/curl/blob/392bff4abf6b177d0770bcb26e43ae5d7addd7c9/lib/multi.c#L1618

We should probably rearrange that code so that it doesn't set the timestamp if it changes the state to `CURLM_STATE_CONNECT_PEND`.

> (and similar for PRETRANSFER_TIME)

That's stored after the "do" phase is complete, which is when the request has been issued so if that's wrong it has a different explanation...

**Comment by bagder** [comment]
@RedDwarf69 wrote:

> I would say the problem arises from the fact curl handles a lot of protocols with an uniform interface.

I really don't think that's the reason. It is possibly the reason why the documentation is a bit vague and unspecific since it covers all protocols, but I don't think that implies this needs to be a problem.

**Comment by lonezor** [comment]
> > Isn't the example above with a value similar to 6.100001 rather than 0.000228 above more in line with what STARTTRANSFER_TIME should describe?
> 
> Yes I would agree, but that's not what the code currently does. `TIMER_STARTSINGLE` is set when it first _considers_ doing the connect and will be stored even if it at that point finds out it needs to queue it and perform it later:

This sounds like a good idea to measure the time when it actually leaves the queue. It's a generic solution that works well for most uses.

As a sidenote, when looking specifically at HTTP/2. Even if we get the perfect timestamp when taking the entry from the curl queue, we also have an outbound queue in the nghttp2 library.

I did a test where _nghttp2_submit_request()_ and _nghttp2_session_send()_ are iterated a large number of times with no delay. The first 100 requests are sent out immediately but the rest are queued. This is seen by monitoring the callback that indicates that a frame is just about to be sent out using _nghttp2_session_callbacks_set_before_frame_send_callback()_:
```
nghttp2_submit_request(), stream id 1
before_frame_send_callback stream id 1 out of total 1)
nghttp2_submit_request(), stream id 3
before_frame_send_callback stream id 3 out of total 2)
nghttp2_submit_request(), stream id 5
before_frame_send_callback stream id 5 out of total 3)
nghttp2_submit_request(), stream id 7
before_frame_send_callback stream id 7 out of total 4)
...
nghttp2_submit_request(), stream id 191
before_frame_send_callback stream id 191 out of total 96)
nghttp2_submit_request(), stream id 193
before_frame_send_callback stream id 193 out of total 97)
nghttp2_submit_request(), stream id 195
before_frame_send_callback stream id 195 out of total 98)
nghttp2_submit_request(), stream id 197
before_frame_send_callback stream id 197 out of total 99)
nghttp2_submit_request(), stream id 199
before_frame_send_callback stream id 199 out of total 100)
nghttp2_submit_request(), stream id 201
nghttp2_submit_request(), stream id 203
nghttp2_submit_request(), stream id 205
nghttp2_submit_request(), stream id 207
nghttp2_submit_request(), stream id 209
nghttp2_submit_request(), stream id 211
nghttp2_submit_request(), stream id 213
...
nghttp2_submit_request(), stream id 19993
nghttp2_submit_request(), stream id 19995
nghttp2_submit_request(), stream id 19997
nghttp2_submit_request(), stream id 19999
before_frame_send_callback stream id 201 out of total 101)
before_frame_send_callback stream id 203 out of total 102)
before_frame_send_callback stream id 205 out of total 103)
before_frame_send_callback stream id 207 out of total 104)
before_frame_send_callback stream id 209 out of total 105)
before_frame_send_callback stream id 211 out of total 106)
...
before_frame_send_callback stream id 19995 out of total 9998)
before_frame_send_callback stream id 19997 out of total 9999)
before_frame_send_callback stream id 19999 out of total 10000)
```
The existence of a queue can be seen in the API description:
https://nghttp2.org/documentation/nghttp2_session_send.html

So if we want the timestamp when actually sending a request, then
_nghttp2_session_callbacks_set_before_frame_send_callback()_
or
_nghttp2_session_callbacks_set_on_frame_send_callback()_

are useful.
